addons/mass_mailing/models/mailing.py

```diff
@@ -1306,36 +1315,117 @@ def _unsubscribe_token(self, res_id, email):
-                modified = True
+                node.attrib['src'] = self._image_to_url(image)
+                did_convert = True
+        # handle the images in mso comments
+        for comment in [c for c in root.iter(lxml.etree.Comment) if mso_re.match(c.text)]:
+            for match in re.findall(r'<img[^>]*src="(data:image/[A-Za-z]+;base64,[^"]*)"', comment.text):
+                did_convert = True
+                image = re.sub(r'data:image/[A-Za-z]+;base64,', '', match).encode()  # base64 image as bytes
+                comment.text = comment.text.replace(match, self._image_to_url(image))
 
-        if modified:
-            return lxml.html.tostring(root, encoding='unicode')
+        return did_convert
 
+    def _crop_vml_images(self, root):
+        """
+        Find VML v:image elements, crop their source images, make an attachement
+        out of them and replace their source with an url to the attachement.
+        """
+        did_crop = False
+        with requests.Session() as session:
+            for comment in [c for c in root.iter(lxml.etree.Comment) if mso_re.match(c.text)]:
+                for match in re.findall(r'<v:image[^>]*>', comment.text):
+                    try:
+                        url = re.search(r'src=\s*\"([^\"]+)\"', match)[1]
+                        # Make sure we have an absolute URL by adding a scheme and host if needed.
+                        absolute_url = url if '//' in url else f"{self.get_base_url()}{url if url.startswith('/') else f'/{url}'}"
+                        image = self._get_image_by_url(absolute_url, session)
+                        if image:
+                            target_width = float(re.search(r'width:\s*([0-9\.]+)\s*px', match)[1])
+                            target_height = float(re.search(r'height:\s*([0-9\.]+)\s*px', match)[1])
+                            image_processor = tools.ImageProcess(image)
+                            image = image_processor.crop_resize(target_width, target_height, 0, 0)
+                            comment.text = comment.text.replace(url, self._image_to_url(base64.b64encode(image.source)))
+                            did_crop = True
+                    except (ValueError, UnidentifiedImageError, AttributeError):
+                        # - ValueError is raised when the url is not a valid url
+                        # - UnidentifiedImageError is raised when the url is not
+                        #   an image
+                        # - AttributeError is raised when one of the regexes
+                        #   fails to match (e.g. when the image has no width or
+                        #   height).
+                        # We choose to ignore these errors so as not to break
+                        # the entire process just for one image's responsive
+                        # cropping behavior).
+                        pass
+        return did_crop
+
+    def _handle_mail_body_images(self, body_html):
```

Comment:
Actually I would merge both '_convert_inline_images_to_urls' and '_handle_mail_body_images' in stable (aka: improving the current '_convert_inline_images_to_urls' to completely do the convert and crop). Reason is that people may call this method in their custom code, and I think it is better to not change the API in table, especially mailing code is present since a long time. People could have custom code, use it in some tools, ... better try to be as much bw-compatible as possible.

Reply:
```diff
diff --git a/addons/mass_mailing/models/mailing.py b/addons/mass_mailing/models/mailing.py
index 4c17356dc7e..b3ec633a4be 100644
--- a/addons/mass_mailing/models/mailing.py
+++ b/addons/mass_mailing/models/mailing.py
@@ -468,7 +468,7 @@ class MassMailing(models.Model):
         ab_testing_cron = self.env.ref('mass_mailing.ir_cron_mass_mailing_ab_testing').sudo()
         for values in vals_list:
             if values.get('body_html'):
-                values['body_html'] = self._handle_mail_body_images(values['body_html'])
+                values['body_html'] = self._convert_inline_images_to_urls(values['body_html'])
             if values.get('ab_testing_schedule_datetime'):
                 at = fields.Datetime.from_string(values['ab_testing_schedule_datetime'])
                 ab_testing_cron._trigger(at=at)
@@ -480,7 +480,7 @@ class MassMailing(models.Model):
 
     def write(self, values):
         if values.get('body_html'):
-            values['body_html'] = self._handle_mail_body_images(values['body_html'])
+            values['body_html'] = self._convert_inline_images_to_urls(values['body_html'])
         # If ab_testing is already enabled on a mailing and the campaign is removed, we raise a ValidationError
         if values.get('campaign_id') is False and any(mailing.ab_testing_enabled for mailing in self) and 'ab_testing_enabled' not in values:
             raise ValidationError(_("A campaign should be set when A/B test is enabled"))
@@ -1369,9 +1369,11 @@ class MassMailing(models.Model):
         """
         Find inline base64 encoded images, make an attachement out of
         them and replace the inline image with an url to the attachement.
+        Find VML v:image elements, crop their source images, make an attachement
+        out of them and replace their source with an url to the attachement.
         """
-        did_convert = False
-        root = lxml.html.fromstring(body_html) if isinstance(body_html, str) else body_html
+        root = lxml.html.fromstring(body_html)
+        did_modify_body = False
         for node in root.iter('img'):
             match = image_re.match(node.attrib.get('src', ''))
             if match:
@@ -1379,16 +1381,8 @@ class MassMailing(models.Model):
                 image = match.group(2).encode()  # base64 image as bytes
 
                 node.attrib['src'] = self._image_to_url(image)
-                did_convert = True
+                did_modify_body = True
 
-        return did_convert
-
-    def _crop_vml_images(self, root):
-        """
-        Find VML v:image elements, crop their source images, make an attachement
-        out of them and replace their source with an url to the attachement.
-        """
-        did_crop = False
         with requests.Session() as session:
             for comment in [c for c in root.iter(lxml.etree.Comment) if mso_re.match(c.text)]:
                 for match in re.findall(r'<v:image[^>]*>', comment.text):
@@ -1403,7 +1397,7 @@ class MassMailing(models.Model):
                             image_processor = tools.ImageProcess(image)
                             image = image_processor.crop_resize(target_width, target_height, 0, 0)
                             comment.text = comment.text.replace(url, self._image_to_url(base64.b64encode(image.source)))
-                            did_crop = True
+                            did_modify_body = True
                     except (ValueError, UnidentifiedImageError, AttributeError):
                         # - ValueError is raised when the url is not a valid url
                         # - UnidentifiedImageError is raised when the url is not
@@ -1415,12 +1409,7 @@ class MassMailing(models.Model):
                         # the entire process just for one image's responsive
                         # cropping behavior).
                         pass
-        return did_crop
-
-    def _handle_mail_body_images(self, body_html):
-        parsed_body = lxml.html.fromstring(body_html)
-        did_convert_images = self._convert_inline_images_to_urls(parsed_body)
-        did_crop_images = self._crop_vml_images(parsed_body)
-        if did_convert_images or did_crop_images:
-            return lxml.html.tostring(parsed_body, encoding='unicode')
+
+        if did_modify_body:
+            return lxml.html.tostring(root, encoding='unicode')
         return body_html
```